[Docs] Add documentation for typeorm (#26535)
yugabyte/yugabyte-db: docs/content/preview/drivers-orms/nodejs/typeorm.md

---
title: TypeORM
headerTitle: Use an ORM
linkTitle: Use an ORM
description: Node.js TypeORM support for YugabyteDB
headcontent: Node.js ORM support for YugabyteDB
aliases:
  - /integrations/typeorm/
menu:
  preview:
    identifier: typeorm-1
    parent: nodejs-drivers
    weight: 600
type: docs
---

<ul class="nav nav-tabs-alt nav-tabs-yb">
  <li >
    <a href="../sequelize/" class="nav-link ">
      <i class="fa-brands fa-node-js" aria-hidden="true"></i>
      Sequelize
    </a>
  </li>
  <li >
    <a href="../prisma/" class="nav-link ">
      <i class="fa-brands fa-node-js" aria-hidden="true"></i>
      Prisma
    </a>
  </li>
  <li >
    <a href="../typeorm/" class="nav-link active">
      <i class="fa-brands fa-node-js" aria-hidden="true"></i>
      TypeORM
    </a>
  </li>
</ul>

> [TypeORM](https://typeorm.io/) is an ORM that can run in NodeJS, Browser, Cordova, PhoneGap, Ionic, React Native, NativeScript, Expo, and Electron platforms and can be used with TypeScript and JavaScript (ES2021). It supports both Active Record and Data Mapper patterns, unlike all other JavaScript ORMs currently in existence, which means you can write high-quality, loosely coupled, scalable, maintainable applications in the most productive way.

Reviewer: ```suggestion
[TypeORM](https://typeorm.io/) is an ORM that can run in NodeJS, Browser, Cordova, PhoneGap, Ionic, React Native, NativeScript, Expo, and Electron platforms, and can be used with TypeScript and JavaScript (ES2021). Unlike other current JavaScript ORMs, it supports both Active Record and Data Mapper patterns, which means you can write high-quality, loosely coupled, scalable, maintainable applications in the most productive way.
```

Because YugabyteDB is PostgreSQL-compatible, TypeORM supports the YugabyteDB YSQL API.

This page provides details for getting started with TypeORM for connecting to YugabyteDB.

> ## Working with domain objects

Reviewer: Not sure what the point of this section is. It doesn't have any content, just a description of what the section is for

This section describes how to use Node.js models (domain objects) to store and retrieve data from a YugabyteDB cluster.

## CRUD operations

The following steps demonstrate how to perform common tasks required for Node.js application development using TypeORM.

### Step 1: Create a Node.js project and install TypeORM package

> Before proceeding with the next steps, you need to have Node.js installed on your machine. Refer to [Downloading and installing Node.js and npm](https://docs.npmjs.com/downloading-and-installing-node-js-and-npm#using-a-node-installer-to-install-node-js-and-npm).

Reviewer: ```suggestion
Before proceeding, you need to install Node.js on your machine. Refer to [Downloading and installing Node.js and npm](https://docs.npmjs.com/downloading-and-installing-node-js-and-npm#using-a-node-installer-to-install-node-js-and-npm) in the npm documentation.
```

> To create a basic Node.js project and install the `TypeORM` package, do the following:

Reviewer: ```suggestion
To create a basic Node.js project and install the `typeorm` package, do the following:
```

1. Create a new directory and initialize a Node.js project. This creates a `package.json` file.

    ```sh
    mkdir nodejs-typeorm-example && cd nodejs-typeorm-example
    npm init -y
    ```

1. Configure typescript by first creating a `tsconfig.json` file.

    ```sh
    npx tsc --init
    ```

1. Update the `tsconfig.json` file.

    ```
    {
      "compilerOptions": {

>         "target": "ES6",

Reviewer 2: There's no need to use an old version of ECMAScript. You can use whatever you would like.

TypeORM currently ships compiled to ES2021 (but that will be increased in the next version), so using something older is not really necessary.

The important thing about tsconfig is:
```json
{
  "experimentalDecorators": true,
  "emitDecoratorMetadata": true
}
```

FYI: We (TypeORM maintainers) are in the processes of rewriting our docs. The docs for the initial setup which were copied here could be re-written to take advantage of more modern JS features (we'll update them in the near future). The current instructions should still work though.

Author: Thanks @alumni for the review, will make the changes accordingly

        "module": "CommonJS",
        "outDir": "./dist",
        "rootDir": "./src",
        "strict": true,
        "experimentalDecorators": true,
        "emitDecoratorMetadata": true
        }
    }
    ```

> 1. Install the typeorm package and its related dependency packages, along with typescript and the pg driver.

Reviewer: ```suggestion
1. Install the `typeorm` package and its related dependency packages, along with typescript and the pg driver.
```
